File: src/main/java/work/stdpi/pge/editor/modmenu/PgeSettingsScreen.java
package work.stdpi.pge.editor.modmenu;

import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.gui.widget.ButtonWidget;
import net.minecraft.client.gui.widget.SliderWidget;
import net.minecraft.text.Text;
import work.stdpi.pge.editor.logic.EditorManager;

public class PgeSettingsScreen extends Screen {

    private final Screen parent;

    public PgeSettingsScreen(Screen parent) {
        super(Text.literal("Paper Engine Editor"));
        this.parent = parent;
    }

    @Override
    protected void init() {
        int centerX = this.width / 2;
        int top = this.height / 2 - 34;

        this.addDrawableChild(
            new FontWeightSlider(centerX - 100, top, 200, 20)
        );
        this.addDrawableChild(
            new GlyphWidthSlider(centerX - 100, top + 24, 200, 20)
        );
        this.addDrawableChild(
            ButtonWidget.builder(Text.literal("Done"), button -> close())
                .dimensions(centerX - 100, top + 52, 200, 20)
                .build()
        );
    }

    @Override
    public void render(
        DrawContext context,
        int mouseX,
        int mouseY,
        float deltaTicks
    ) {
        this.renderBackground(context, mouseX, mouseY, deltaTicks);
        super.render(context, mouseX, mouseY, deltaTicks);
        context.drawCenteredTextWithShadow(
            this.textRenderer,
            this.title,
            this.width / 2,
            this.height / 2 - 48,
            0xFFFFFF
        );
        context.drawCenteredTextWithShadow(
            this.textRenderer,
            Text.literal(
                "Embedded Intel One Mono with persistent terminal tuning."
            ),
            this.width / 2,
            this.height / 2 - 34,
            0xA0A0A0
        );
    }

    @Override
    public void close() {
        MinecraftClient.getInstance().setScreen(parent);
    }

    private static final class FontWeightSlider extends SliderWidget {

        private FontWeightSlider(int x, int y, int width, int height) {
            super(
                x,
                y,
                width,
                height,
                Text.empty(),
                toValue(EditorManager.INSTANCE.getTerminalFontWeight())
            );
            updateMessage();
        }

        @Override
        protected void updateMessage() {
            EditorManager.TerminalFontWeight weight = getFontWeight();
            this.setMessage(Text.literal("Font Weight: " + formatWeight(weight)));
        }

        @Override
        protected void applyValue() {
            EditorManager.INSTANCE.setTerminalFontWeight(getFontWeight());
        }

        private EditorManager.TerminalFontWeight getFontWeight() {
            EditorManager.TerminalFontWeight[] weights = EditorManager.TerminalFontWeight.values();
            int index = (int) Math.round(this.value * (weights.length - 1));
            return weights[Math.max(0, Math.min(weights.length - 1, index))];
        }

        private static double toValue(EditorManager.TerminalFontWeight weight) {
            EditorManager.TerminalFontWeight[] weights = EditorManager.TerminalFontWeight.values();
            return (double) weight.ordinal() / (double) (weights.length - 1);
        }

        private static String formatWeight(EditorManager.TerminalFontWeight weight) {
            return switch (weight) {
                case LIGHT -> "Light";
                case REGULAR -> "Regular";
                case MEDIUM -> "Medium";
                case BOLD -> "Bold";
            };
        }
    }

    private static final class GlyphWidthSlider extends SliderWidget {

        private static final int MIN_WIDTH = 1;
        private static final int MAX_WIDTH = 24;

        private GlyphWidthSlider(int x, int y, int width, int height) {
            super(
                x,
                y,
                width,
                height,
                Text.empty(),
                toValue(EditorManager.INSTANCE.getTerminalCellWidthPx())
            );
            updateMessage();
        }

        @Override
        protected void updateMessage() {
            int pixels = getWidthPixels();
            this.setMessage(Text.literal("Terminal Zoom: " + pixels + " px cell width"));
        }

        @Override
        protected void applyValue() {
            EditorManager.INSTANCE.setTerminalCellWidthPx(getWidthPixels());
        }

        private int getWidthPixels() {
            return (
                MIN_WIDTH +
                (int) Math.round(this.value * (MAX_WIDTH - MIN_WIDTH))
            );
        }

        private static double toValue(int width) {
            return (
                (double) (width - MIN_WIDTH) / (double) (MAX_WIDTH - MIN_WIDTH)
            );
        }
    }

}

File: src/main/kotlin/work/stdpi/pge/editor/modmenu/PgeSettingsScreen.kt
package work.stdpi.pge.editor.modmenu

import net.minecraft.client.MinecraftClient
import net.minecraft.client.gui.DrawContext
import net.minecraft.client.gui.screen.Screen
import net.minecraft.client.gui.widget.ButtonWidget
import net.minecraft.client.gui.widget.ButtonWidget.PressAction
import net.minecraft.client.gui.widget.SliderWidget
import net.minecraft.text.Text
import work.stdpi.pge.editor.logic.EditorManager
import work.stdpi.pge.editor.logic.EditorManager.TerminalFontWeight
import work.stdpi.pge.editor.logic.EditorManager.TerminalSupersample
import kotlin.math.max
import kotlin.math.min

class PgeSettingsScreen(private val parent: Screen?) : Screen(Text.literal("Paper Engine Editor")) {
  override fun init() {
    val centerX = this.width / 2
    val top = this.height / 2 - 46

    this.addDrawableChild<FontWeightSlider?>(FontWeightSlider(centerX - 100, top, 200, 20))
    this.addDrawableChild<SupersampleSlider?>(SupersampleSlider(centerX - 100, top + 24, 200, 20))
    this.addDrawableChild<GlyphWidthSlider?>(GlyphWidthSlider(centerX - 100, top + 48, 200, 20))
    this.addDrawableChild<ButtonWidget?>(
        ButtonWidget.builder(Text.literal("Done"), PressAction { button: ButtonWidget? -> close() })
            .dimensions(centerX - 100, top + 76, 200, 20)
            .build())
  }

  override fun render(context: DrawContext, mouseX: Int, mouseY: Int, deltaTicks: Float) {
    this.renderBackground(context, mouseX, mouseY, deltaTicks)
    super.render(context, mouseX, mouseY, deltaTicks)
    context.drawCenteredTextWithShadow(
        this.textRenderer, this.title, this.width / 2, this.height / 2 - 48, 0xFFFFFF)
    context.drawCenteredTextWithShadow(
        this.textRenderer,
        Text.literal("Embedded Intel One Mono with persistent terminal tuning."),
        this.width / 2,
        this.height / 2 - 46,
        0xA0A0A0)
  }

  override fun close() {
    MinecraftClient.getInstance().setScreen(parent)
  }

  private class FontWeightSlider(x: Int, y: Int, width: Int, height: Int) :
      SliderWidget(
          x,
          y,
          width,
          height,
          Text.empty(),
          toValue(EditorManager.terminalFontWeight)) {
    init {
      updateMessage()
    }

    override fun updateMessage() {
      val weight = this.fontWeight
      this.setMessage(Text.literal("Font Weight: " + formatWeight(weight)))
    }

    override fun applyValue() {
      EditorManager.terminalFontWeight = this.fontWeight
    }

    val fontWeight: TerminalFontWeight
      get() {
        val weights: Array<TerminalFontWeight> = TerminalFontWeight.entries.toTypedArray()
        val index = Math.round(this.value * (weights.size - 1)).toInt()
        return weights[max(0, min(weights.size - 1, index))]
      }

    companion object {
      private fun toValue(weight: TerminalFontWeight): Double {
        val weights: Array<TerminalFontWeight> = TerminalFontWeight.entries.toTypedArray()
        return weight.ordinal.toDouble() / (weights.size - 1).toDouble()
      }

      private fun formatWeight(weight: TerminalFontWeight): String {
        return when (weight) {
          TerminalFontWeight.LIGHT -> "Light"
          TerminalFontWeight.REGULAR -> "Regular"
          TerminalFontWeight.MEDIUM -> "Medium"
          TerminalFontWeight.BOLD -> "Bold"
        }
      }
    }
  }

  private class GlyphWidthSlider(x: Int, y: Int, width: Int, height: Int) :
      SliderWidget(
          x,
          y,
          width,
          height,
          Text.empty(),
          toValue(EditorManager.terminalCellWidthPx)) {
    init {
      updateMessage()
    }

    override fun updateMessage() {
      val pixels = this.widthPixels
      this.setMessage(Text.literal("Terminal Zoom: " + pixels + " px cell width"))
    }

    override fun applyValue() {
      EditorManager.terminalCellWidthPx = this.widthPixels
    }

    val widthPixels: Int
      get() = (MIN_WIDTH + Math.round(this.value * (MAX_WIDTH - MIN_WIDTH)).toInt())

    companion object {
      private const val MIN_WIDTH = 1
      private const val MAX_WIDTH = 24

      private fun toValue(width: Int): Double {
        return ((width - MIN_WIDTH).toDouble() / (MAX_WIDTH - MIN_WIDTH).toDouble())
      }
    }
  }

  private class SupersampleSlider(x: Int, y: Int, width: Int, height: Int) :
      SliderWidget(
          x,
          y,
          width,
          height,
          Text.empty(),
          toValue(EditorManager.terminalSupersample)) {
    init {
      updateMessage()
    }

    override fun updateMessage() {
      val supersample = this.supersample
      this.setMessage(Text.literal("Supersample: " + supersample.value + "x"))
    }

    override fun applyValue() {
      EditorManager.terminalSupersample = this.supersample
    }

    val supersample: TerminalSupersample
      get() {
        val values: Array<TerminalSupersample> = TerminalSupersample.entries.toTypedArray()
        val index = Math.round(this.value * (values.size - 1)).toInt()
        return values[max(0, min(values.size - 1, index))]
      }

    companion object {
      private fun toValue(supersample: TerminalSupersample): Double {
        val values: Array<TerminalSupersample> = TerminalSupersample.entries.toTypedArray()
        return supersample.ordinal.toDouble() / (values.size - 1).toDouble()
      }
    }
  }
}
